# ML-04 — Search Intelligence Data Contract

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Beshair-Khan/flyrank_ml_internship/blob/main/work/notebooks/w03_data_contract.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

In [ ]:
import os
import pandas as pd
from dotenv import load_dotenv
from huggingface_hub import hf_hub_download

load_dotenv()
hf_token = os.getenv('HF_TOKEN')
print("Token added successfully:", hf_token is not None)
file_c = hf_hub_download(
    repo_id="FlyRank/internship-warehouse",
    filename="dim_content.parquet",
    repo_type="dataset",
    token=hf_token)
file_f = hf_hub_download(
    repo_id="FlyRank/internship-warehouse",
    filename="fact_content_daily_performance/month=2026-03/data_0.parquet",
    repo_type="dataset",
    token=hf_token)
df_c = pd.read_parquet(file_c)
df_f = pd.read_parquet(file_f)
print(df_c.shape, df_f.shape)
df_c.head()


Token added successfully: True
(519606, 26) (9841378, 30)


,client_hash_id,content_hash_id,keyword_hash_id,url_hash_id,keyword_char_count,keyword_token_count,url_char_count,content_created_date,content_updated_date,content_type,search_volume,competition,competition_level,cpc,main_intent,backlinks,category_count,keyword_created_date,provider_used,model_used,char_count,word_count,last_optimized_date,optimization_eligible_date,is_published,is_deleted
0,client_04660893ae39614a,content_004de9653278b5a4,keyword_e754999ab88dd9f2,url_d6091f18cf628794,22,4,108,2026-05-30,2026-07-01,keyword article,30.0,0.91,HIGH,0.98,transactional,16.0,3,2026-05-12,gemini-generate-content,gemini-3-flash-preview,15682.0,2555.0,None,None,True,False
1,client_04660893ae39614a,content_00dc5efae381b2ab,keyword_4329d7aede8e208b,url_3a66d2f2e36823ca,31,6,95,2026-06-12,2026-07-01,keyword article,10.0,0.00,LOW,0.00,commercial,0.0,4,2026-06-01,gemini-generate-content,gemini-3-flash-preview,15438.0,2430.0,None,None,True,False
2,client_04660893ae39614a,content_01410f2556c327ac,keyword_9b08047d3d2a0406,url_809eda7a7e20b3b2,22,5,82,2026-05-09,2026-07-01,keyword article,480.0,0.36,MEDIUM,0.62,informational,169.0,4,2026-05-06,gemini-generate-content,gemini-3-flash-preview,16576.0,2645.0,None,None,True,False
3,client_04660893ae39614a,content_019f27f634053ca7,keyword_e7cec7ab1804c1c2,url_5fb42bafc4399861,14,3,92,2026-06-15,2026-06-15,keyword article,0.0,0.00,LOW,0.00,transactional,0.0,4,2026-06-01,gemini-generate-content,gemini-3-flash-preview,15457.0,2522.0,None,None,True,False
4,client_04660893ae39614a,content_01efa71faea45dcc,keyword_56b0062a1d8b7524,url_ece0abc3e5fb75f9,24,6,98,2026-05-21,2026-06-01,keyword article,2400.0,0.70,HIGH,0.90,transactional,52.0,4,2026-05-12,gemini-generate-content,gemini-3-flash-preview,15776.0,2552.0,None,None,True,False


In [4]:
df_f.head()

,report_date,client_hash_id,content_hash_id,client_has_gsc,client_has_ga4,gsc_data_available,ga4_data_available,gsc_impressions,gsc_clicks,gsc_sum_position,gsc_avg_position,ga4_pageviews,ga4_sessions,ga4_users,ga4_engaged_sessions,ga4_total_engagement_sec,sessions_organic,sessions_direct,sessions_referral,sessions_social,sessions_paid,sessions_ai,ai_chatgpt,ai_perplexity,ai_gemini,ai_copilot,ai_claude,ai_meta,ai_other,scroll_events
0,2026-03-01,client_73cda7b4e4f265ea,content_b7e512995f79d5a6,True,False,True,None,20,0,67,3.350000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2026-03-01,client_73cda7b4e4f265ea,content_05597932fe4da067,True,False,True,None,1,0,0,0.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2026-03-01,client_73cda7b4e4f265ea,content_7a105f548d9c6916,True,False,True,None,125,1,616,4.928000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,2026-03-01,client_73cda7b4e4f265ea,content_905aa32a0230694e,True,False,True,None,7,0,28,4.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,2026-03-01,client_73cda7b4e4f265ea,content_a3ea9792f793ec72,True,False,True,None,11,0,25,2.272727,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


## 1. Unit of analysis + time window

*One row = one what, over which dates? State it, then verify it below.*

* **Unit Of Analysis:** 1 row = 1 unique webpage (content_hash_id), aggregated over March 2026.
* **Time Window:** month=2026-03, loaded directly from the warehouse partition (verified below)

In [5]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
page_stats=df_f.groupby('content_hash_id').agg(
    total_clicks=('gsc_clicks','sum'),
    total_impressions=('gsc_impressions', 'sum'),
    avg_position=('gsc_sum_position', 'mean')
).reset_index()
page_stats

,content_hash_id,total_clicks,total_impressions,avg_position
0,content_000005d4ced12088,0,86,199.967742
1,content_00001e488b74b799,0,0,0.000000
2,content_00007bd2985b77c3,0,47,8.032258
3,content_00008950670cb6b5,0,0,0.000000
4,content_0000a348850eb1fc,0,0,0.000000
...,...,...,...,...
331432,content_ffffbc148e416f89,0,1,3.129032
331433,content_ffffc282ab2cbe62,0,0,0.000000
331434,content_ffffc58385523096,37,2482,293.645161
331435,content_ffffe701567e982c,0,0,0.000000


In [6]:
# Time Window Claim
df_f['report_date']=pd.to_datetime(df_f['report_date'])
max_date=df_f['report_date'].max()
min_date=df_f['report_date'].min()
print(f"The max date is {max_date} & the min date is {min_date}")

The max date is 2026-03-31 00:00:00 & the min date is 2026-03-01 00:00:00


## 2. Fields: feature / label / context / excluded

*Sort every field you plan to touch into these four buckets. Excluded needs a why.

* **Features:** word_count, days_since_update, total_clicks, total_impressions, avg_position (avg_position only used in the leakage demo, otherwise excluded (see below)).
* **Label:** opportunity_flags (built from is_declining, is_page_one, is_stale)
* **Context:** content_hash_id (page identifier, not predictive).
* **Excluded:** avg_position (leaks is_page_one), is_deleted (may be a consequence of decline, not an independent signal), provider_used/model_used (flagged as a future feature to test, excluded for now)

In [7]:
df_c['content_updated_date'] = pd.to_datetime(df_c['content_updated_date'])
kept_rows = df_c[df_c['content_updated_date'] <= max_date].copy()
kept_rows['days_since_update'] = (max_date - kept_rows['content_updated_date']).dt.days
features = ['content_hash_id', 'content_updated_date', 'word_count', 'content_type', 'is_deleted', 'days_since_update']
kept_rows[features]
kept_rows=kept_rows[kept_rows['is_deleted']==False]
kept_rows[features]
len(kept_rows)

42821

In [8]:
df=kept_rows[features].merge(page_stats,on='content_hash_id',how='inner')
df.head()

,content_hash_id,content_updated_date,word_count,content_type,is_deleted,days_since_update,total_clicks,total_impressions,avg_position
0,content_04c67f3541177192,2026-02-25,3168.0,keyword article,False,34,2,331,153.516129
1,content_05acc92c165f4386,2026-02-25,4135.0,keyword article,False,34,0,33,9.967742
2,content_0b33d8960857ad90,2026-02-25,3232.0,keyword article,False,34,0,0,0.000000
3,content_0f30e04e709c7b5d,2026-02-25,3211.0,keyword article,False,34,0,145,38.000000
4,content_167472cd0802a8f3,2026-02-25,3149.0,keyword article,False,34,0,232,89.516129


In [9]:
df['is_page_one'] = (df['avg_position'] <= 10).astype(int)
df['is_stale'] = (df['days_since_update'] >= 180).astype(int)
df['opportunity_flags'] = df['is_page_one'] + df['is_stale']
df['opportunity_flags'].value_counts().sort_index()

opportunity_flags
0    21199
1    12624
2     3558
Name: count, dtype: int64

## 3. Verify it with queries (grain, counts, missing values, windows)

*Every claim above gets a query cell here. A contract claim without a query next to it is a guess.*

Time Window: This starter snapshot (content_refresh_anonymized.csv) has no date/month column — it's a single static cross-section, not time-partitioned. A real time window (e.g. month=2026-03) will apply once we move to the Hugging Face warehouse release in Week 3.

* **Grain Check:** Confirming zero duplicate content_id entries.
* **Row Count & Completeness:** Measuring overall row count and non-null availability across features.
* **Availability Filter (IS TRUE check):** Verifying how many rows survive when requiring valid, non-null feature values.

In [10]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
dup=df['content_hash_id'].duplicated().sum()
print(f"There are {dup} duplicates in dataframe")
shape=df.shape
print(f"There are {shape[0]} rows and {shape[1]} columns in dataset")
print(df.notnull().sum())

There are 0 duplicates in dataframe
There are 37381 rows and 12 columns in dataset
content_hash_id         37381
content_updated_date    37381
word_count              24799
content_type            37381
is_deleted              37381
days_since_update       37381
total_clicks            37381
total_impressions       37381
avg_position            37381
is_page_one             37381
is_stale                37381
opportunity_flags       37381
dtype: int64


In [11]:
df['word_count']=df['word_count'].fillna(df['word_count'].median())
df.isnull().sum()

content_hash_id         0
content_updated_date    0
word_count              0
content_type            0
is_deleted              0
days_since_update       0
total_clicks            0
total_impressions       0
avg_position            0
is_page_one             0
is_stale                0
opportunity_flags       0
dtype: int64

In [12]:
available = df.notnull().all(axis=1)
print(f"{available.sum()} out of {len(df)} rows have complete data")

37381 out of 37381 rows have complete data


## 4. Data limits

*What can this data never tell you? Unbalanced history, GSC-only early rows, window overlaps.*

**What can this data never tell you?**

This slice drops from roughly 519,600 raw pages to 37,381 after removing future dated updates and deleted pages, and since only one month of data is available the label captures page one position and staleness only, with no true declining trend signal like earlier notebooks used it also still can't explain why a page underperforms or confirm that a refresh would actually help, since that requires a real experiment, not just historical data.

In [18]:
import numpy as np
def precision_at_k(scores, labels, k):
    order = np.argsort(-np.asarray(scores))
    topk = np.asarray(labels)[order[:k]]
    return topk.mean()

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
feat=['word_count','days_since_update',	'total_clicks','total_impressions']
X=df[feat]
y=(df['opportunity_flags']>=1).astype(int)
X_train, X_test, y_train,y_test=train_test_split(X,y,test_size=0.25,random_state=42)
model=DecisionTreeClassifier(random_state=42)
model.fit(X_train, y_train)
prediction=model.predict(X_test)
test_scores = model.predict_proba(X_test)[:, 1]
print(precision_at_k(test_scores, y_test, 50))

0.96


In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
feat=['word_count','days_since_update',	'total_clicks','total_impressions','avg_position']
X=df[feat]
y=(df['opportunity_flags']>=1).astype(int)
X_train, X_test, y_train,y_test=train_test_split(X,y,test_size=0.25,random_state=42)
model=DecisionTreeClassifier(random_state=42)
model.fit(X_train, y_train)
prediction=model.predict(X_test)
test_scores = model.predict_proba(X_test)[:, 1]
print(precision_at_k(test_scores, y_test, 50))

1.0


Adding `avg_position` as a feature made `Precision@50` jump from `0.96` to a `perfect 1.0`. This is not real improvement it's leakage. `is_page_one` (half of the label `opportunity_flags`) is defined directly as `avg_position <= 10`, so giving the model `avg_position` is close to handing it the answer outright. A perfect score is a red flag, not a win: it means the model found a shortcut instead of learning a real pattern from the honest features `(word_count, days_since_update, total_clicks, total_impressions)`.

## Self-check

Before you submit, confirm each line honestly:

- [ Yes ] Every section above is filled — markdown thinking AND the code that backs it
- [ Yes ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ Yes ] No client names, URLs, or private queries anywhere
- [ Yes ] My claims use careful words: observed, measured, directional, decision-support
- [ Yes ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.